File: Lab 6/queue.py
def main():
    queue = []  # Initialize an empty queue
    
    while True:
        print("\nQueue Operations:")
        print("1. Enqueue")
        print("2. Dequeue")
        print("3. Display")
        print("4. Exit")
        
        choice = int(input("Enter your choice: "))

        if choice == 1:  # Enqueue
            value = int(input("Enter the value to enqueue: "))
            queue.append(value)
            print(f"{value} enqueued to the queue.")
        
        elif choice == 2:  # Dequeue
            if queue:
                print(f"Dequeued value: {queue.pop(0)}")
            else:
                print("Queue is empty! Cannot dequeue.")
        
        elif choice == 3:  # Display
            if queue:
                print("Queue elements are:", *queue)
                print(f"Front: {queue[0]}, Rear: {queue[-1]}")
            else:
                print("Queue is empty!")
        
        elif choice == 4:  # Exit
            print("Exiting...")
            break
        
        else:
            print("Invalid choice! Try again.")
        
        # Show current state of the queue after each operation
        if queue:
            print("Current queue:", *queue)
        else:
            print("Queue is empty.")    

if __name__ == "__main__":
    main()

File: Lab 6/queue.c
#include <stdio.h>
#define MAX 5 // Maximum size of the queue

int queue[MAX], front = -1, rear = -1;

int main() {
    int choice, value;

    while (1) {
        printf("\nQueue Operations:\n");
        printf("1. Enqueue\n");
        printf("2. Dequeue\n");
        printf("3. Display\n");
        printf("4. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1: // Enqueue
                if (rear == MAX - 1) {
                    printf("Queue is full! Cannot enqueue.\n");
                } else {
                    if (front == -1) { // Set front if queue is initially empty
                        front = 0;
                    }
                    printf("Enter the value to enqueue: ");
                    scanf("%d", &value);
                    rear++;
                    queue[rear] = value;
                    printf("%d enqueued to the queue.\n", value);
                    printf("Current front: %d, rear: %d\n", front, rear);
                }
                break;

            case 2: // Dequeue
                if (front == -1 || front > rear) {
                    printf("Queue is empty! Cannot dequeue.\n");
                } else {
                    printf("Dequeued value: %d\n", queue[front]);
                    front++;
                    printf("Current front: %d, rear: %d\n", front, rear);
                    if (front > rear) { // Reset the queue when all elements are dequeued
                        front = rear = -1;
                        printf("Queue is now empty.\n");
                    }
                }
                break;

            case 3: // Display
                if (front == -1 || front > rear) {
                    printf("Queue is empty!\n");
                } else {
                    printf("Queue elements are: ");
                    for (int i = front; i <= rear; i++) {
                        printf("%d ", queue[i]);
                    }
                    printf("\n");
                }
                break;

            case 4: // Exit
                printf("Exiting...\n");
                return 0;

            default:
                printf("Invalid choice! Try again.\n");
        }

        //EXTRA:> Show current state of the queue after each operation
        if (front == -1 || front > rear) {
            printf("Queue is empty.\n");
        } else {
            printf("Current queue: ");
            for (int i = front; i <= rear; i++) {
                printf("%d ", queue[i]);
            }
            printf("\n");
        }
    }

    return 0;
}
